File: C#/EntityFramework_ORM/BLL/Services/CategoryService.cs
using BLL.Interfaces;
using DAL.Entities;
using DAL.Repositories;

namespace BLL.Services;

public class CategoryService : IService<Category, int>
{
    private readonly CategoryRepository _repository;

    public CategoryService(CategoryRepository repository)
    {
        _repository = repository;
    }

    public Category Add(Category entity) => _repository.Add(entity);
    
    public Category? Delete(int id) => _repository.Delete(id);

    public List<Category> GetAll() => _repository.GetAll();

    public Category? GetById(int id) => _repository.GetById(id);

    public Category Update(Category entity) => _repository.Update(entity);
}

File: Go/GORM_ORM/bll/services/categoryService.go
package services

import (
	"GORM_ORM/dal/entities"
	"GORM_ORM/dal/repositories"
)

type CategoryService interface {
	GetAll() ([]entities.Category, error)
	GetByID(id int) (*entities.Category, error)
	Create(category *entities.Category) error
	Update(category *entities.Category) error
	Delete(id int) error
}

type categoryService struct {
	repo repositories.CategoryRepository
}

func NewCategoryService(repo repositories.CategoryRepository) CategoryService {
	return &categoryService{repo: repo}
}

func (s *categoryService) GetAll() ([]entities.Category, error) {
	return s.repo.GetAll()
}

func (s *categoryService) GetByID(id int) (*entities.Category, error) {
	return s.repo.GetByID(id)
}

func (s *categoryService) Create(category *entities.Category) error {
	return s.repo.Create(category)
}

func (s *categoryService) Update(category *entities.Category) error {
	return s.repo.Update(category)
}

func (s *categoryService) Delete(id int) error {
	return s.repo.Delete(id)
}
